### ISSUE TITLE
C++ modules not handling internal linkage

### ISSUE THREAD

**Issue by rnikander** [llvm/llvm-project#90259]
In the following example code, I have two variables `int abc` that should be private to the files (internal linkage). They are in anonymous namespaces. But, they conflict with each other.

`main.cc`:
```
import mymod;
int main() {
    foo();
}
```
`mymod.cppm`:
```
module;
#include <stdio.h>
export module mymod;
export import :part1;
namespace { int abc; }
export void foo() {
    printf("In foo() %d\n", abc);
}
```
`part1.cppm`:
```
export module mymod:part1;
namespace { int abc; }
export void part1_fun() {}
```

Compiler error:

```
mymod.cppm:9:29: error: reference to 'abc' is ambiguous
    9 |     printf("In foo() %d\n", abc);
      |                             ^
/.../part1.cppm:4:17: note: candidate found by name lookup is '(anonymous namespace)::abc'
    4 | namespace { int abc; }
      |                 ^
mymod.cppm:6:17: note: candidate found by name lookup is '(anonymous namespace)::abc'
    6 | namespace { int abc; }
      |                 ^
```

**Comment by rnikander** [comment]
This is worse than I thought. Even when I move the anonymous namespaces into _implementation_ units, the names still conflict.

**Comment by ChuanqiXu9** [comment]
Can you try "-fexperimental-modules-reduced-bmi"?

**Comment by rnikander** [comment]
> -fexperimental-modules-reduced-bmi

How new is this flag? I built my clang on March 28 and it's not recognizing it. (I use 2-phase compilation usually.)

**Comment by rnikander** [comment]
> Can you try "-fexperimental-modules-reduced-bmi"?

Mm... I just built Clang from main branch and I get: `clang++: error: unknown argument: '-fexperimental-modules-reduced-bmi'` Do I need to build Clang a certain way or do something to enable that flag? (By the way I've been using 2-phase compilation.)

**Comment by ChuanqiXu9** [comment]
> > Can you try "-fexperimental-modules-reduced-bmi"?
> 
> Mm... I just built Clang from main branch and I get: `clang++: error: unknown argument: '-fexperimental-modules-reduced-bmi'` Do I need to build Clang a certain way or do something to enable that flag? (By the way I've been using 2-phase compilation.)

I guess there are some configuration mismatches. I run it with my modules project every day: https://github.com/alibaba/async_simple/blob/CXX20Modules/CMakeLists.txt#L36-L38

BTW, if you're using 2-phase compilation, the build system may need to be changed slightly. https://clang.llvm.org/docs/StandardCPlusPlusModules.html#reduced-bmi

But you can verify if reduced BMI can solve the problem.

**Comment by rnikander** [comment]
@ChuanqiXu9 Anything stand out as incorrect in this command line?

```
/Users/rob/Dev/llvm-project/build_phase1/bin/clang++ -std=c++2b 
   -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk 
   -nostdinc++ -nostdlib++ -isystem /Users/rob/Dev/llvm-project/build_phase2_x86_macos/include/c++/v1 
   -fexperimental-modules-reduced-bmi --precompile part1.cppm -o part1.pcm -fmodule-output=part1.reduced.pcm
clang++: error: unknown argument: '-fexperimental-modules-reduced-bmi'
```

**Comment by ChuanqiXu9** [comment]
> @ChuanqiXu9 Anything stand out as incorrect in this command line?
> 
> ```
> /Users/rob/Dev/llvm-project/build_phase1/bin/clang++ -std=c++2b 
>    -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk 
>    -nostdinc++ -nostdlib++ -isystem /Users/rob/Dev/llvm-project/build_phase2_x86_macos/include/c++/v1 
>    -fexperimental-modules-reduced-bmi --precompile part1.cppm -o part1.pcm -fmodule-output=part1.reduced.pcm
> clang++: error: unknown argument: '-fexperimental-modules-reduced-bmi'
> ```

It looks good. I just checked godbolt, which can accept '-fexperimental-modules-reduced-bmi' too: https://godbolt.org/z/brE341jGs

So I still believe there might be some configuration mismatches in your build. Possibly you're using old clang.

**Comment by rnikander** [comment]
Yup, sorry, I had screwed up my `git` commands. Now I have compiled latest Clang and tried it. It does not seem to make a difference. Things that should have internal linkage in different TUs are conflicting.

```
clang++ -std=c++2b ... -fexperimental-modules-reduced-bmi --precompile part1.cppm 
   -o part1.pcm -fmodule-output=part1.reduced.pcm
clang++ -std=c++2b ...  -fexperimental-modules-reduced-bmi --precompile mymod.cppm 
   -o mymod.pcm -fmodule-output=mymod.reduced.pcm -fmodule-file=mymod:part1=part1.reduced.pcm

mymod.cppm:9:29: error: reference to 'abc' is ambiguous
    9 |     printf("In foo() %d\n", abc);
      |                             ^
part1.cppm:4:17: note: candidate found by name lookup is '(anonymous namespace)::abc'
    4 | namespace { int abc; }
      |                 ^
mymod.cppm:6:17: note: candidate found by name lookup is '(anonymous namespace)::abc'
    6 | namespace { int abc; }
      |                 ^
```

**Comment by ChuanqiXu9** [comment]
oh, there might be some mismatches. I guess it might work if it is static variable. I'll handle the anonymous namespace case.
